mongo_orchestration/common.py

```diff
@@ -15,14 +15,84 @@
 # limitations under the License.
 
 import collections
+import copy
 import json
 import os
-import copy
+import stat
+import tempfile
 
 DEFAULT_BIND = os.environ.get('MO_HOST', 'localhost')
 DEFAULT_PORT = int(os.environ.get('MO_PORT', '8889'))
 DEFAULT_SERVER = 'cherrypy'
 
+# Username for included client x509 certificate.
+DEFAULT_SUBJECT = (
+    'C=US,ST=New York,L=New York City,O=MongoDB,OU=KernelUser,CN=client'
+)
+DEFAULT_CLIENT_CERT = os.path.join(
+    os.environ.get(
+        'MONGO_ORCHESTRATION_HOME', os.path.dirname(__file__)),
+    'lib',
+    'client.pem'
+)
+
+
+class BaseModel(object):
+    """Base object for Server, ReplicaSet, and ShardedCluster."""
+
+    _user_roles = [
+        {'role': 'userAdminAnyDatabase', 'db': 'admin'},
+        {'role': 'clusterAdmin', 'db': 'admin'},
+        {'role': 'dbAdminAnyDatabase', 'db': 'admin'},
+        {'role': 'readWriteAnyDatabase', 'db': 'admin'}
+    ]
+
+    @property
+    def key_file(self):
+        """Get the path to the key file containig our auth key, or None."""
+        if self.auth_key:
+            key_file_path = os.path.join(tempfile.mkdtemp(), 'key')
+            with open(key_file_path, 'w') as fd:
+                fd.write(self.auth_key)
+            os.chmod(key_file_path, stat.S_IRUSR)
```

Comment:
tempfile.mkdtemp makes the file readable and writable by the user. The chmod call makes it read-only.